### PR TITLE
ws-proto: shell module

### PR DESCRIPTION
**Pull request by merlin-northern**
ws-protocol shell module

ChangeLog:none
Signed-off-by: Peter Grzybowski <peter@northern.tech>

### CODE REVIEW

File: ws/shell/model.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+	SessionId string `json:"session_id" msgpack:"session_id"`
+	//message status, currently normal and error message types are supported
+	Status MenderShellMessageStatus `json:"status_code" msgpack:"status_code"`
+	//the message payload, if
+	// * .Type===MessageTypeShellCommand interpreted as keystrokes and passed
+	//   to the stdin of the terminal running the shell.
+	// * .Type===MessageTypeSpawnShell interpreted as user_id and passed
+	//   to the session.NewMenderShellSession.
+	Data []byte `json:"data" msgpack:"data"`
+}
```
Comment (alfrunes) [inline ws/shell/model.go:49]: This definition seems superfluous. It's completely overlapping with `ProtoMsg`. Why not just declare a `ProtoMsg` struct instead?

File: ws/shell/model.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package shell
+
+import (
+	"github.com/mendersoftware/go-lib-micro/ws"
+)
+
+var (
```
Comment (alfrunes) [inline ws/shell/model.go:21]:
```suggestion
const (
```

File: ws/shell/model.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	MessageTypeShellCommand = "shell"
+	MessageTypeSpawnShell   = "new"
+	MessageTypeStopShell    = "stop"
+)
+
+type MenderShellMessageStatus int
+
+const (
+	NormalMessage MenderShellMessageStatus = iota
```
Comment (alfrunes) [inline ws/shell/model.go:30]:
```suggestion
	NormalMessage MenderShellMessageStatus = iota + 1
```
Starting on `1` will catch errors due to uninitialized structs upon validation.

File: ws/shell/model.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	// * .Type===MessageTypeShellCommand interpreted as keystrokes and passed
+	//   to the stdin of the terminal running the shell.
+	// * .Type===MessageTypeSpawnShell interpreted as user_id and passed
+	//   to the session.NewMenderShellSession.
+	Data []byte `json:"data" msgpack:"data"`
+}
+
+func (s *MenderShellMessage) UnMarshal(m *ws.ProtoMsg) {
+	s.Data = m.Body
+}
```
Comment (alfrunes) [inline ws/shell/model.go:53]: This seems redundant.

File: ws/shell/model.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+		Header: ws.ProtoHdr{
+			Proto:     ws.ProtoTypeShell,
+			MsgType:   s.Type,
+			SessionID: s.SessionId,
+			Properties: map[string]interface{}{
+				"status": s.Status,
+			},
+		},
+		Body: s.Data,
+	}
```
Comment (alfrunes) [inline ws/shell/model.go:70]: Why not use this declaration in the first place?

**Review by alfrunes — COMMENTED** [review]
I don't understand the need for `MenderShellMessage`. To me it seems like an alias for `ProtoMsg`. If we don't have any models for the message `Body` I don't think we need anything else than the constant definitions here.

File: ws/shell/model.go (hunk 1 above)
Comment (merlin-northern) [reply ws/shell/model.go:49]: the reason: I wanted to minimize the required modifications to mender-shell (MenderShellMessage used in few places) and deviceconnect.

**Comment by merlin-northern** [comment]
discussed privately (I lost of course ;-))

**Review by alfrunes — APPROVED** [review]

**Pull request merged**